File: src/main/java/com/br/service/DynamoDbService.java
package com.br.service;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
import software.amazon.awssdk.services.dynamodb.model.PutItemRequest;

import java.time.Instant;
import java.util.Map;

@ApplicationScoped
public class DynamoDbService {

    @Inject
    DynamoDbClient dynamoDb;

    @ConfigProperty(name = "app.dynamodb-table")
    String tableName;

    /**
     * Persists a processed image record. The imageId (hash key) is the
     * deterministic output S3 key, making this call naturally idempotent —
     * re-running with the same inputs simply overwrites the same item.
     */
    public void saveProcessedRecord(
            String outputKey,
            String dealerId,
            String carroId,
            String sourceKey,
            String size,
            String format) {

        dynamoDb.putItem(PutItemRequest.builder()
            .tableName(tableName)
            .item(Map.of(
                "imageId",    AttributeValue.fromS(outputKey),
                "dealerId",   AttributeValue.fromS(dealerId),
                "carroId",    AttributeValue.fromS(carroId),
                "sourceKey",  AttributeValue.fromS(sourceKey),
                "size",       AttributeValue.fromS(size),
                "format",     AttributeValue.fromS(format),
                "status",     AttributeValue.fromS("PROCESSED"),
                "processedAt", AttributeValue.fromS(Instant.now().toString())
            ))
            .build());
    }
}

File: app/services/dynamodb_service.py
import os
from datetime import datetime, timezone
import boto3


class DynamoDbService:
    def __init__(self) -> None:
        self._table_name = os.environ["DYNAMODB_TABLE"]
        self._client = boto3.client("dynamodb")

    def save_processed_record(
        self,
        output_key: str,
        dealer_id: str,
        vehicle_id: str,
        source_key: str,
        size: str,
        fmt: str,
    ) -> None:
        """Persists a processed image record.
        The imageId (hash key) is the deterministic output S3 key,
        making this call naturally idempotent — same inputs overwrites the same item.
        """
        self._client.put_item(
            TableName=self._table_name,
            Item={
                "imageId":     {"S": output_key},
                "dealerId":    {"S": dealer_id},
                "vehicleId":     {"S": vehicle_id},
                "sourceKey":   {"S": source_key},
                "size":        {"S": size},
                "format":      {"S": fmt},
                "status":      {"S": "PROCESSED"},
                "processedAt": {"S": datetime.now(timezone.utc).isoformat()},
            },
        )
